app/src/main/java/com/epfl/drawyourpath/mainpage/fragments/FriendsFragment.kt

```diff
@@ -66,4 +71,21 @@ class FriendsFragment : Fragment(R.layout.fragment_friends) {
             }
         })
     }
+
+    private fun onScanQRClicked() {
+        val mainActivity = requireActivity() as MainActivity
+        mainActivity.scanQRCode()
+            .thenApply {
+                // TODO: Add friend from ID "it"
+                if (it == null) {
+                    Toast.makeText(mainActivity, "Scan cancelled", Toast.LENGTH_LONG).show()
+                }
+                else {
+                    Toast.makeText(mainActivity, "Scanned $it", Toast.LENGTH_LONG).show()
+                }
+            }
```

Comment:
Here I don't see a way for `it` to be anything but `null.` Looking at the `scanQRCode()` function, it seems that it returns either a `CompletableFuture` with `null` as value, or completes exceptionally. Am I missing something?

Reply:
The flow is a bit complex due to how Android handles messages through activities.

The flow is as followed:
1. User calls `scanQRCode()`.
2. `scanQRCode()` calls `ActivityCompat.requestPermissions()`.
3. Permission result is returned through `MainActivity.onRequestPermissionsResult()` which is part of the Android SDK.
4. `MainActivity.onRequestPermissionsResult()` calls `launchFriendQRScanner()` which starts the `QRScannerActivity`, expecting a result.
5. The user scans a QR code and the `QRScannerActivity` activity sets the result with `setResult()` and terminates to give back execution to the `MainActivity` with `finish()`.
6. `MainActivity.onActivityResult()` is called as part of the Android SDK for activity completion.
7. `MainActivity.onActivityResult()` gets the result of the scan through an intent and fires the future.

Reply:
I think I understand it a bit better with your nice explaination but still, I feel like the result is contained in `qrScanResult` and not in `result`, which is what `scanQRCode()` returns. The line 86 of `MainActivity` seems to be "inverted" to me... but there is maybe still something I don't get.

Reply:
We store the new future for later completion. We return the future without it being completed because we need to wait for results before firing it.

Reply:
Yes I think I can see the logic now, even though it's quite confusing, thank you for the explaination!